src/frogger.cpp

```diff
@@ -31,40 +31,83 @@ int get_hiv_steps_per_year(SEXP r_hiv_steps_per_year) {
   return hiv_steps_per_year;
 }
 
-leapfrog::TensorMap1<int> get_age_groups_hiv_span(const Rcpp::List projection_parameters,
-                                                  std::string hiv_age_stratification) {
+int get_age_groups_hiv(const std::string hiv_age_stratification) {
   int age_groups_hiv;
-  SEXP data;
   if (hiv_age_stratification == "full") {
     age_groups_hiv = 66;
-    data = projection_parameters["hAG_SPAN_full"];
-  } else if (hiv_age_stratification == "coarse") {
-    age_groups_hiv = 9;
-    data = projection_parameters["hAG_SPAN_coarse"];
   } else {
-    Rcpp::stop(
-        "Invalid HIV age stratification must be 'full' or 'coarse' got '%s'.",
-        hiv_age_stratification);
+    // We've already validated that hiv_age_stratification is one of "full" or "coarse"
+    age_groups_hiv = 9;
   }
-  if (LENGTH(data) != age_groups_hiv) {
-    Rcpp::stop("Invalid size of data, expected %d got %d", age_groups_hiv,
-               LENGTH(data));
+  return age_groups_hiv;
+}
+
+template<typename... Args>
+auto parse_data_int(const Rcpp::List data, const std::string& key, Args... dims) {
+  constexpr std::size_t rank = sizeof...(dims);
+  Eigen::array<int, rank> dimensions{ static_cast<int>(dims)... };
+
+  int product = 1;
+  for (size_t i = 0; i < rank; ++i) {
+    product *= dimensions[i];
   }
```

Comment:
```suggestion
  const int product = std::accumulate(dimensions.begin(), dimensions.end(), 1, std::multiplies<size_t>());
```

See https://en.cppreference.com/w/cpp/algorithm/accumulate

There are _some_ nice bits in C++ :)